File: src/main/java/pl/laskowski/marcin/model/Field.java
package pl.laskowski.marcin.model;

/**
 * Created by Marcin Laskowski.
 */

public class Field {

    private static final int EMPTY_FIELD = 0;

    private final Point position;
    private int value = EMPTY_FIELD;

    public Field(int x, int y) {
        this.position = new Point(x, y);
    }

    public int value() {
        return value;
    }

    public void set(int value) {
        this.value = value;
    }

    public boolean isEmpty() {
        return value == EMPTY_FIELD;
    }

    public int getX() {
        return position.x();
    }

    public int getY() {
        return position.y();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Field field = (Field) o;

        if (value != field.value) return false;
        return position.equals(field.position);
    }

    @Override
    public int hashCode() {
        int result = position.hashCode();
        result = 31 * result + value;
        return result;
    }

    public void clear() {
        value = EMPTY_FIELD;
    }

    public Field copy() {
        Field field = new Field(position.x(), position.y());
        field.value = this.value;
        return field;
    }

    public boolean haveSamePosition(Field field) {
        return this.position.equals(field.position);
    }

    public Point position() {
        return position;
    }
}

File: app/src/main/java/pl/laskowski/marcin/model/Field.kt
package pl.laskowski.marcin.model

data class Field(
    val position: Point,
    var value: Int = EMPTY_FIELD
) {

    constructor(x: Int, y: Int): this(Point(x, y))

    fun value() = value

    fun set(value: Int) {
        this.value = value
    }

    val isEmpty: Boolean
        get() = value == EMPTY_FIELD

    val x = position.x
    val y = position.y

    fun clear() {
        value = EMPTY_FIELD
    }

    fun haveSamePosition(field: Field): Boolean {
        return position == field.position
    }

    fun position() = position

    companion object {
        private const val EMPTY_FIELD = 0
    }
}